DOCS: Added PlayerInput.ActionEvent documentation

lyndon-unity opened the pull request:
### Description

 Added PlayerInput.ActionEvent documentation
Updated PlayerInput documentation to improve, adding missing parameters, adding some examples, moving some see also links into the remarks as part of explanations. 

### Testing status & QA

Reviewed generated docs using the DocTools package and checked the validation results

### Overall Product Risks

- Complexity: low
- Halo Effect: low

### Comments to reviewers

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.
    - Jira port for the next release set as "Resolved".
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [X] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

lyndon-unity committed:
Documented PlayerInput.ActionEvent

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +1980,54 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
+        /// <remarks>
+        /// Represents an event invoked in response to actions being triggered.
+        ///
+        /// Contains the Id and name of the action being triggered and the associated UnityAction to handle the action response.
```
ID?

ekcoh replied:
Is UnityAction a type that should be referenced?

lyndon-unity replied:
Good points - fixing

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +1980,54 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
+        /// Contains the Id and name of the action being triggered and the associated UnityAction to handle the action response.
+        /// </remarks>
+        /// <example>
+        /// <code>
```
Empty code example, looks unintentional?

lyndon-unity replied:
Removing - was unintentional - no example makes sense here as this is done internally.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +1980,54 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
             }
 
+            /// <summary>
+            /// Construct an action event and associated it with an action.
```
Typo, review language

lyndon-unity replied:
Changing 
associated -> associate

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +1980,54 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
+            /// The event will be associated with the specified action. The action must be part of an action asset.
+            /// </remarks>
+            /// <param name="action">The action to associate with the event. The action must be part of an action asset.</param>
+            /// <exception cref="ArgumentNullException">The action is null.</exception>
```
I believe null should have <c> formatting here?

ekcoh replied:
Should this have <seealso cref="InputAction" />?

lyndon-unity replied:
Ill add < c > around the null. I assume these whats meant.

I don't understand the '?' query

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2004,6 +2041,14 @@ public ActionEvent(InputAction action)
+            /// The event will be associated with the specified action. 
+            /// </remarks>
+            /// <param name="actionGUID">Action GUID</param>
+            /// <param name="name">name of the action</param>
```
Capitalise "name" to "Name"?

lyndon-unity replied:
Done (in the "Name of action" part)

ekcoh commented:
CI errors on docs formatting it seems

lyndon-unity committed:
Fixed PlayerInput link

(PlayerInputEditor is internal)

lyndon-unity committed:
Code review feedback and auto formatting fix

lyndon-unity committed:
Further updates to improve the documentation page score

duckets commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +1980,67 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
+        ///
+        /// <example>
+        /// <code>
+        /// public class MyPlayerScript : MonoBehaviour
```
Does this example compile? I can't see any "using" statements

duckets replied:
I have added them.

duckets committed:
added "using" statements to code example

lyndon-unity committed:
Updated PlayerInput documentation to fix documentation errors

Extended summaries, added missing parameters, added some examples, removed non recommended documentation fields (moving some into remarks as recommended)

lyndon-unity committed:
Merge branch 'docs-quality-week-2024-playerinput-actionevent' of github.com:Unity-Technologies/InputSystem into docs-quality-week-2024-playerinput-actionevent

lyndon-unity committed:
Added missing "using" statements

lyndon-unity commented:
> CI errors on docs formatting it seems

Fixed formatting, addressed a whole load more documentation issues from the low score on the documentation assessment.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2023,7 +2243,9 @@ public class DeviceLostEvent : UnityEvent<PlayerInput>
         /// </summary>
-        /// <seealso cref="deviceRegainedEvent"/>
+        /// <remarks>
+        /// The device regained event can be set on via <see cref="deviceRegainedEvent"/>
```
Missing full stop

lyndon-unity replied:
Fixing in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2032,7 +2254,9 @@ public class DeviceRegainedEvent : UnityEvent<PlayerInput>
         /// </summary>
-        /// <seealso cref="controlsChangedEvent"/>
+        /// <remarks>
+        /// The changed event can be set on via <see cref="controlsChangedEvent"/>
```
Missing full stop

lyndon-unity replied:
Fixing in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2014,7 +2232,9 @@ public ActionEvent(Guid actionGUID, string name = null)
         /// </summary>
-        /// <seealso cref="deviceLostEvent"/>
+        /// <remarks>
+        /// The device lost event can be set on via <see cref="deviceLostEvent"/>
```
Missing full stop

lyndon-unity replied:
Fixing in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2004,6 +2213,15 @@ public ActionEvent(InputAction action)
+            /// <remarks>
+            /// The event will be associated with the specified action.
+            /// </remarks>
+            /// <param name="actionGUID">Action GUID</param>
```
Missing full stop

lyndon-unity replied:
Fixing in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -2004,6 +2213,15 @@ public ActionEvent(InputAction action)
+            /// The event will be associated with the specified action.
+            /// </remarks>
+            /// <param name="actionGUID">Action GUID</param>
+            /// <param name="name">Name of the action</param>
```
Missing full stop

lyndon-unity replied:
Fixing in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1978,19 +2139,67 @@ private void SwitchControlSchemeInternal(ref InputControlScheme controlScheme, p
+        /// </code>
+        /// </example>
+        /// </remarks>
+        /// <seealso cref="PlayerInput.actionEvents"/>
```
Should we add a seals here for InputAction.CallbackContext as well since that what the sample shows?

lyndon-unity replied:
Adding in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1763,6 +1905,25 @@ private void OnDisable()
+        /// using UnityEngine.InputSystem;
+        /// // Component to sit next to PlayerInput.
+        /// [RequireComponent(typeof(PlayerInput))]
+        /// public class MyPlayerLogic : MonoBehaviour
```
Enabling and disabling this component multiple times would add one new each handler on every OnEnable? Is this intended

lyndon-unity replied:
Nope I'll add an OnDisable function too.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1068,14 +1204,20 @@ public static PlayerInput Instantiate(GameObject prefab, int playerIndex = -1, s
+        /// <param name="controlScheme">Control scheme to activate.</param>
+        /// <param name="splitScreenIndex">Which split screen to instantiate on.</param>
+        /// <param name="pairWithDevices">Which devices to limit pairing to.</param>
+        /// <returns>Newly created PlayerInput component</returns>
```
Missing full stop

lyndon-unity replied:
Thank you. I'll pay more attention to these. Adding in next commit

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -1038,13 +1167,20 @@ public static PlayerInput FindFirstPairedToDevice(InputDevice device)
         /// that PlayerInput will automatically pair with available, unpaired devices based on the control schemes (if any)
         /// present in <see cref="actions"/> or on the bindings therein (if no control schemes are present).</param>
-        /// <returns></returns>
+        /// <returns>Newly created PlayerInput component</returns>
```
Missing full stop

ekcoh replied:
<c> formatting missing it seems

ekcoh replied:
Might add ". Never <c>null</c>." since the user then knows they do not need to null check. In lack of nullable syntax.

lyndon-unity replied:
Fixing full stop and formatting in next commit
It looks like it could be null so I won't add that part.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -989,11 +1110,18 @@ public void SwitchCurrentActionMap(string mapNameOrId)
+        /// <remarks>
+        /// <example>
+        /// <code>
+        /// PlayerInput player = PlayerInput.GetPlayerByIndex(0);
```
Are we ok with non-compilable examples? @duckets

lyndon-unity replied:
It should compile but not in isolation. I'll await doc writer feedback

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/PlayerInput/PlayerInput.cs:
```diff
@@ -961,6 +1068,20 @@ public void SwitchCurrentControlScheme(string controlScheme, params InputDevice[
+        ///
+        /// <example>
+        /// <code>
+        /// PlayerInput.all[0].SwitchCurrentActionMap("Player");
```
Similar comment as previously on non-compilable example

lyndon-unity replied:
It should compile but not in isolation. I'll await doc writer feedback

ekcoh commented:
Still has similar errors in CI as my branch (See dry-run target):

PVP-150-1: UnityEngine.InputSystem.PlayerInput: onActionTriggered: <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: onDeviceLost: <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: onDeviceRegained: <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: onControlsChanged: <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: TDevice GetDevice(): <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: void ActivateInput(): <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: void DeactivateInput(): <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: void SwitchCurrentActionMap(string): <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: PlayerInput GetPlayerByIndex(int): <example> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.PlayerInput: PlayerInput Instantiate(GameObject, int, string, int, InputDevice): <example> in block context (only allowed in top-level context)
(and 3 more errors)

lyndon-unity commented:
> Still has similar errors in CI as my branch (See dry-run target):

I've fixed these. Several were already there but I've tided it up.

lyndon-unity committed:
Docs feedback, including fixed 'examples' tag to be top level (rather than nested in 'remarks')

lyndon-unity committed:
Fixed mixed block error in documentation

lyndon-unity committed:
Adding missing word

lyndon-unity committed:
Merge branch 'develop' into docs-quality-week-2024-playerinput-actionevent
